File: server/ai-tools.ts
import { GoogleGenerativeAI } from "@google/generative-ai";
import * as stats from "simple-statistics";
import Papa from "papaparse";
import * as XLSX from "xlsx";
import fs from "fs/promises";

interface DatasetInfo {
  shape: [number, number];
  columns: string[];
  dtypes: Record<string, string>;
  head: any[];
}

interface EDAResult {
  analysis_type: string;
  dataset_info: DatasetInfo;
  statistics: Record<string, any>;
  missing_values: Record<string, number>;
  correlation?: Record<string, any>;
  visualizations?: Array<{
    type: string;
    title: string;
    data: any;
  }>;
}

interface ForecastResult {
  target_column: string;
  method: string;
  periods: number;
  data_points: number;
  forecast_values: any[];
  visualization?: {
    type: string;
    title: string;
    data: any;
  };
}

// Global dataset storage
let currentDataset: Record<string, any[]> | null = null;
let datasetColumns: string[] = [];

/**
 * Detect column data type
 */
function detectType(values: any[]): string {
  const nonNullValues = values.filter(v => v !== null && v !== undefined && v !== "");
  if (nonNullValues.length === 0) return "string";
  
  const sample = nonNullValues[0];
  if (typeof sample === "number") return "number";
  if (typeof sample === "boolean") return "boolean";
  if (!isNaN(Number(sample))) return "number";
  return "string";
}

/**
 * Load dataset from file path
 */
export async function loadDataset(filePath: string): Promise<{ success: boolean; error?: string; info?: DatasetInfo }> {
  try {
    const fileContent = await fs.readFile(filePath);
    let data: any[] = [];
    
    // Check file extension
    if (filePath.endsWith(".csv")) {
      // Parse CSV
      const parsed = Papa.parse(fileContent.toString("utf-8"), {
        header: true,
        dynamicTyping: true,
        skipEmptyLines: true,
      });

      if (parsed.errors.length > 0) {
        return { success: false, error: `CSV parsing error: ${parsed.errors[0].message}` };
      }

      data = parsed.data;
    } else if (filePath.endsWith(".xlsx") || filePath.endsWith(".xls")) {
      // Parse Excel
      const workbook = XLSX.read(fileContent, { type: "buffer" });
      const sheetName = workbook.SheetNames[0];
      const worksheet = workbook.Sheets[sheetName];
      data = XLSX.utils.sheet_to_json(worksheet);
    } else {
      return { success: false, error: "Unsupported file format" };
    }

    if (data.length === 0) {
      return { success: false, error: "No data found in file" };
    }

    // Convert to column-based storage
    const columns = Object.keys(data[0]);
    const dataset: Record<string, any[]> = {};
    
    columns.forEach(col => {
      dataset[col] = data.map(row => row[col]);
    });

    currentDataset = dataset;
    datasetColumns = columns;

    // Get data types
    const dtypes: Record<string, string> = {};
    columns.forEach(col => {
      dtypes[col] = detectType(dataset[col]);
    });

    return {
      success: true,
      info: {
        shape: [data.length, columns.length],
        columns,
        dtypes,
        head: data.slice(0, 5),
      },
    };
  } catch (error: any) {
    return { success: false, error: error.message };
  }
}

/**
 * Get numeric columns
 */
function getNumericColumns(): string[] {
  if (!currentDataset) return [];
  
  return datasetColumns.filter(col => {
    const values = currentDataset![col].filter(v => v !== null && v !== undefined && v !== "");
    return values.length > 0 && !isNaN(Number(values[0]));
  });
}

/**
 * Calculate correlation between two arrays
 */
function correlation(x: number[], y: number[]): number {
  if (x.length !== y.length || x.length === 0) return 0;
  
  const meanX = stats.mean(x);
  const meanY = stats.mean(y);
  const stdX = stats.standardDeviation(x);
  const stdY = stats.standardDeviation(y);
  
  if (stdX === 0 || stdY === 0) return 0;
  
  let sum = 0;
  for (let i = 0; i < x.length; i++) {
    sum += ((x[i] - meanX) / stdX) * ((y[i] - meanY) / stdY);
  }
  
  return sum / (x.length - 1);
}

/**
 * Perform Exploratory Data Analysis
 */
export async function performEDA(analysisType: string = "comprehensive"): Promise<EDAResult | { error: string }> {
  if (!currentDataset) {
    return { error: "No dataset loaded. Please load a dataset first." };
  }

  try {
    const rowCount = currentDataset[datasetColumns[0]].length;
    const dtypes: Record<string, string> = {};
    
    datasetColumns.forEach(col => {
      dtypes[col] = detectType(currentDataset![col]);
    });

    // Get first 5 rows
    const head: any[] = [];
    for (let i = 0; i < Math.min(5, rowCount); i++) {
      const row: any = {};
      datasetColumns.forEach(col => {
        row[col] = currentDataset![col][i];
      });
      head.push(row);
    }

    const result: EDAResult = {
      analysis_type: analysisType,
      dataset_info: {
        shape: [rowCount, datasetColumns.length],
        columns: datasetColumns,
        dtypes,
        head,
      },
      statistics: {},
      missing_values: {},
      visualizations: [],
    };

    // Get statistics for numeric columns
    const numericColumns = getNumericColumns();

    numericColumns.forEach((col) => {
      const values = currentDataset![col]
        .filter(v => v !== null && v !== undefined && v !== "")
        .map(v => Number(v));
      
      if (values.length > 0) {
        result.statistics[col] = {
          count: values.length,
          mean: stats.mean(values),
          std: stats.standardDeviation(values),
          min: stats.min(values),
          max: stats.max(values),
          median: stats.median(values),
          q1: stats.quantile(values, 0.25),
          q3: stats.quantile(values, 0.75),
        };
      }
    });

    // Missing values
    datasetColumns.forEach((col) => {
      const nullCount = currentDataset![col].filter(v => v === null || v === undefined || v === "").length;
      if (nullCount > 0) {
        result.missing_values[col] = nullCount;
      }
    });

    // Correlation matrix for numeric columns
    if (numericColumns.length > 1) {
      const correlationData: Record<string, any> = {};
      
      numericColumns.forEach((col1) => {
        correlationData[col1] = {};
        const values1 = currentDataset![col1]
          .filter(v => v !== null && v !== undefined && v !== "")
          .map(v => Number(v));
        
        numericColumns.forEach((col2) => {
          if (col1 === col2) {
            correlationData[col1][col2] = 1.0;
          } else {
            const values2 = currentDataset![col2]
              .filter(v => v !== null && v !== undefined && v !== "")
              .map(v => Number(v));
            
            correlationData[col1][col2] = correlation(values1, values2);
          }
        });
      });
      
      result.correlation = correlationData;

      // Create heatmap data
      result.visualizations?.push({
        type: "heatmap",
        title: "Correlation Heatmap",
        data: {
          z: numericColumns.map(col1 => 
            numericColumns.map(col2 => correlationData[col1][col2])
          ),
          x: numericColumns,
          y: numericColumns,
          type: "heatmap",
          colorscale: "RdBu",
          zmid: 0,
        },
      });
    }

    // Distribution data for numeric columns (first 5)
    if (analysisType === "comprehensive" || analysisType === "distribution") {
      numericColumns.slice(0, 5).forEach((col) => {
        const values = currentDataset![col]
          .filter(v => v !== null && v !== undefined && v !== "")
          .map(v => Number(v));
        
        // Create histogram data
        result.visualizations?.push({
          type: "histogram",
          title: `Distribution of ${col}`,
          data: {
            x: values,
            type: "histogram",
            name: col,
            nbinsx: 30,
          },
        });

        // Create box plot data
        result.visualizations?.push({
          type: "box",
          title: `Box Plot of ${col}`,
          data: {
            y: values,
            type: "box",
            name: col,
          },
        });
      });
    }

    return result;
  } catch (error: any) {
    return { error: `EDA failed: ${error.message}` };
  }
}

/**
 * Exponential smoothing forecast
 */
function exponentialSmoothing(data: number[], periods: number, alpha: number = 0.3): number[] {
  const forecast: number[] = [];
  let lastValue = data[data.length - 1];
  let lastSmoothed = data[data.length - 1];
  
  for (let i = 0; i < periods; i++) {
    const smoothed = alpha * lastValue + (1 - alpha) * lastSmoothed;
    forecast.push(smoothed);
    lastValue = smoothed;
    lastSmoothed = smoothed;
  }
  
  return forecast;
}

/**
 * Perform time series forecasting
 */
export async function forecastData(
  targetColumn: string,
  periods: number = 30,
  method: string = "auto"
): Promise<ForecastResult | { error: string }> {
  if (!currentDataset) {
    return { error: "No dataset loaded. Please load a dataset first." };
  }

  if (!datasetColumns.includes(targetColumn)) {
    return { error: `Column '${targetColumn}' not found in dataset.` };
  }

  try {
    const values = currentDataset[targetColumn]
      .filter(v => v !== null && v !== undefined && v !== "")
      .map(v => Number(v));

    if (values.length < 10) {
      return { error: "Not enough data points for forecasting. Need at least 10 data points." };
    }

    // Use exponential smoothing as default method
    const forecastValues = exponentialSmoothing(values, periods);

    const result: ForecastResult = {
      target_column: targetColumn,
      method: "Exponential Smoothing",
      periods,
      data_points: values.length,
      forecast_values: forecastValues.map((val, idx) => ({
        index: idx + 1,
        value: val,
      })),
      visualization: {
        type: "forecast",
        title: `Forecast: ${targetColumn}`,
        data: {
          historical: {
            x: Array.from({ length: values.length }, (_, i) => i),
            y: values,
            type: "scatter",
            mode: "lines",
            name: "Historical Data",
          },
          forecast: {
            x: Array.from({ length: periods }, (_, i) => values.length + i),
            y: forecastValues,
            type: "scatter",
            mode: "lines",
            name: "Forecast",
            line: { dash: "dash", color: "red" },
          },
        },
      },
    };

    return result;
  } catch (error: any) {
    return { error: `Forecasting failed: ${error.message}` };
  }
}

/**
 * Get column information
 */
export async function getColumnInfo(): Promise<{ columns: Record<string, any>; total_rows: number; total_columns: number } | { error: string }> {
  if (!currentDataset) {
    return { error: "No dataset loaded." };
  }

  try {
    const columnInfo: Record<string, any> = {};
    const rowCount = currentDataset[datasetColumns[0]].length;

    datasetColumns.forEach((col) => {
      const values = currentDataset![col];
      const nonNullValues = values.filter(v => v !== null && v !== undefined && v !== "");
      
      columnInfo[col] = {
        dtype: detectType(values),
        non_null_count: nonNullValues.length,
        null_count: values.length - nonNullValues.length,
        unique_values: new Set(values).size,
        sample_values: values.slice(0, 3),
      };
    });

    return {
      columns: columnInfo,
      total_rows: rowCount,
      total_columns: datasetColumns.length,
    };
  } catch (error: any) {
    return { error: error.message };
  }
}

/**
 * AI Agent with function calling
 */
export async function createAIAgent(apiKey: string, modelName: string = "gemini-2.0-flash-exp") {
  const genAI = new GoogleGenerativeAI(apiKey);
  
  const tools = [
    {
      name: "perform_eda",
      description: "Perform Exploratory Data Analysis on the loaded dataset. Returns statistical summaries, correlation matrices, and visualizations.",
      parameters: {
        type: "object",
        properties: {
          analysis_type: {
            type: "string",
            enum: ["comprehensive", "statistical", "correlation", "distribution"],
            description: "Type of analysis to perform",
          },
        },
      },
    },
    {
      name: "forecast_data",
      description: "Perform time series forecasting on a specific column. Returns forecast values and visualization.",
      parameters: {
        type: "object",
        properties: {
          target_column: {
            type: "string",
            description: "Name of the column to forecast",
          },
          periods: {
            type: "number",
            description: "Number of periods to forecast (default: 30)",
          },
          method: {
            type: "string",
            enum: ["auto", "exponential", "moving_average"],
            description: "Forecasting method to use",
          },
        },
        required: ["target_column"],
      },
    },
    {
      name: "get_column_info",
      description: "Get detailed information about all columns in the dataset, including data types, null counts, and sample values.",
      parameters: {
        type: "object",
        properties: {},
      },
    },
  ];

  const model = genAI.getGenerativeModel({
    model: modelName,
    tools: [{ functionDeclarations: tools as any }],
  });

  return { model, tools };
}

/**
 * Execute function call
 */
export async function executeFunctionCall(functionName: string, args: any): Promise<any> {
  switch (functionName) {
    case "perform_eda":
      return await performEDA(args.analysis_type || "comprehensive");
    case "forecast_data":
      return await forecastData(args.target_column, args.periods || 30, args.method || "auto");
    case "get_column_info":
      return await getColumnInfo();
    default:
      return { error: `Unknown function: ${functionName}` };
  }
}

File: server/ai_tools.py
#!/usr/bin/env python3
"""
AI Analysis Tools using LangChain and Gemini
Provides EDA and forecasting capabilities for datasets
"""

import sys
import json
import pandas as pd
import matplotlib
matplotlib.use('Agg')  # Non-interactive backend
import matplotlib.pyplot as plt
import seaborn as sns
import io
import base64
from typing import Dict, Any, List
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain.agents import AgentExecutor, create_tool_calling_agent
from langchain.tools import tool
from langchain_core.prompts import ChatPromptTemplate
from prophet import Prophet
from statsmodels.tsa.arima.model import ARIMA
import warnings
warnings.filterwarnings('ignore')

# Global dataset storage
current_dataset: pd.DataFrame = None
dataset_path: str = None

def load_dataset(file_path: str) -> Dict[str, Any]:
    """Load dataset from file path"""
    global current_dataset, dataset_path
    try:
        if file_path.endswith('.csv'):
            current_dataset = pd.read_csv(file_path)
        elif file_path.endswith(('.xlsx', '.xls')):
            current_dataset = pd.read_excel(file_path)
        else:
            return {"error": "Unsupported file format. Please use CSV or Excel files."}
        
        dataset_path = file_path
        
        return {
            "success": True,
            "shape": current_dataset.shape,
            "columns": list(current_dataset.columns),
            "dtypes": {col: str(dtype) for col, dtype in current_dataset.dtypes.items()},
            "head": current_dataset.head().to_dict()
        }
    except Exception as e:
        return {"error": str(e)}

@tool
def perform_eda(analysis_type: str = "comprehensive") -> str:
    """
    Perform Exploratory Data Analysis on the loaded dataset.
    
    Args:
        analysis_type: Type of analysis - 'comprehensive', 'statistical', 'correlation', or 'distribution'
    
    Returns:
        JSON string with analysis results and visualizations
    """
    global current_dataset
    
    if current_dataset is None:
        return json.dumps({"error": "No dataset loaded. Please load a dataset first."})
    
    try:
        results = {
            "analysis_type": analysis_type,
            "dataset_info": {
                "shape": current_dataset.shape,
                "columns": list(current_dataset.columns),
                "dtypes": {col: str(dtype) for col, dtype in current_dataset.dtypes.items()}
            }
        }
        
        # Basic statistics
        results["statistics"] = current_dataset.describe().to_dict()
        
        # Missing values
        missing = current_dataset.isnull().sum()
        results["missing_values"] = {col: int(val) for col, val in missing.items() if val > 0}
        
        # Correlation matrix for numeric columns
        numeric_cols = current_dataset.select_dtypes(include=['number']).columns
        if len(numeric_cols) > 1:
            corr_matrix = current_dataset[numeric_cols].corr()
            results["correlation"] = corr_matrix.to_dict()
            
            # Generate correlation heatmap
            plt.figure(figsize=(10, 8))
            sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', center=0, fmt='.2f')
            plt.title('Correlation Heatmap')
            plt.tight_layout()
            
            buf = io.BytesIO()
            plt.savefig(buf, format='png', dpi=100, bbox_inches='tight')
            buf.seek(0)
            img_base64 = base64.b64encode(buf.read()).decode('utf-8')
            plt.close()
            
            results["visualizations"] = results.get("visualizations", [])
            results["visualizations"].append({
                "type": "heatmap",
                "title": "Correlation Heatmap",
                "image": f"data:image/png;base64,{img_base64}"
            })
        
        # Distribution plots for numeric columns
        if analysis_type in ["comprehensive", "distribution"]:
            for col in numeric_cols[:5]:  # Limit to first 5 numeric columns
                plt.figure(figsize=(10, 6))
                plt.subplot(1, 2, 1)
                current_dataset[col].hist(bins=30, edgecolor='black')
                plt.title(f'Distribution of {col}')
                plt.xlabel(col)
                plt.ylabel('Frequency')
                
                plt.subplot(1, 2, 2)
                current_dataset.boxplot(column=col)
                plt.title(f'Box Plot of {col}')
                
                plt.tight_layout()
                
                buf = io.BytesIO()
                plt.savefig(buf, format='png', dpi=100, bbox_inches='tight')
                buf.seek(0)
                img_base64 = base64.b64encode(buf.read()).decode('utf-8')
                plt.close()
                
                results["visualizations"] = results.get("visualizations", [])
                results["visualizations"].append({
                    "type": "distribution",
                    "column": col,
                    "title": f"Distribution Analysis: {col}",
                    "image": f"data:image/png;base64,{img_base64}"
                })
        
        return json.dumps(results, indent=2)
        
    except Exception as e:
        return json.dumps({"error": f"EDA failed: {str(e)}"})

@tool
def forecast_data(target_column: str, periods: int = 30, method: str = "auto") -> str:
    """
    Perform time series forecasting on the specified column.
    
    Args:
        target_column: Name of the column to forecast
        periods: Number of periods to forecast (default: 30)
        method: Forecasting method - 'auto', 'prophet', or 'arima'
    
    Returns:
        JSON string with forecast results and visualization
    """
    global current_dataset
    
    if current_dataset is None:
        return json.dumps({"error": "No dataset loaded. Please load a dataset first."})
    
    if target_column not in current_dataset.columns:
        return json.dumps({"error": f"Column '{target_column}' not found in dataset."})
    
    try:
        # Prepare data
        data = current_dataset[[target_column]].copy()
        data = data.dropna()
        
        if len(data) < 10:
            return json.dumps({"error": "Not enough data points for forecasting. Need at least 10 data points."})
        
        # Add time index if not present
        data['ds'] = pd.date_range(start='2020-01-01', periods=len(data), freq='D')
        data['y'] = data[target_column]
        
        results = {
            "target_column": target_column,
            "method": method,
            "periods": periods,
            "data_points": len(data)
        }
        
        # Choose forecasting method
        if method == "auto" or method == "prophet":
            # Prophet forecasting
            model = Prophet(yearly_seasonality=True, weekly_seasonality=True, daily_seasonality=False)
            model.fit(data[['ds', 'y']])
            
            future = model.make_future_dataframe(periods=periods)
            forecast = model.predict(future)
            
            results["forecast_values"] = forecast[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail(periods).to_dict('records')
            results["method_used"] = "Prophet"
            
            # Generate forecast plot
            fig = model.plot(forecast)
            plt.title(f'Forecast for {target_column} using Prophet')
            plt.xlabel('Date')
            plt.ylabel(target_column)
            plt.tight_layout()
            
        else:  # ARIMA
            # Simple ARIMA model
            model = ARIMA(data['y'], order=(1, 1, 1))
            fitted_model = model.fit()
            
            forecast_result = fitted_model.forecast(steps=periods)
            
            results["forecast_values"] = [
                {"index": i, "value": float(val)} 
                for i, val in enumerate(forecast_result, 1)
            ]
            results["method_used"] = "ARIMA"
            
            # Generate forecast plot
            plt.figure(figsize=(12, 6))
            plt.plot(data['y'].values, label='Historical Data')
            forecast_index = range(len(data), len(data) + periods)
            plt.plot(forecast_index, forecast_result, label='Forecast', color='red')
            plt.title(f'Forecast for {target_column} using ARIMA')
            plt.xlabel('Time Index')
            plt.ylabel(target_column)
            plt.legend()
            plt.tight_layout()
        
        # Save plot
        buf = io.BytesIO()
        plt.savefig(buf, format='png', dpi=100, bbox_inches='tight')
        buf.seek(0)
        img_base64 = base64.b64encode(buf.read()).decode('utf-8')
        plt.close()
        
        results["visualization"] = {
            "type": "forecast",
            "title": f"Forecast: {target_column}",
            "image": f"data:image/png;base64,{img_base64}"
        }
        
        return json.dumps(results, indent=2)
        
    except Exception as e:
        return json.dumps({"error": f"Forecasting failed: {str(e)}"})

@tool
def get_column_info() -> str:
    """
    Get information about all columns in the dataset.
    
    Returns:
        JSON string with column names, types, and sample values
    """
    global current_dataset
    
    if current_dataset is None:
        return json.dumps({"error": "No dataset loaded."})
    
    try:
        column_info = {}
        for col in current_dataset.columns:
            column_info[col] = {
                "dtype": str(current_dataset[col].dtype),
                "non_null_count": int(current_dataset[col].count()),
                "null_count": int(current_dataset[col].isnull().sum()),
                "unique_values": int(current_dataset[col].nunique()),
                "sample_values": current_dataset[col].head(3).tolist()
            }
        
        return json.dumps({
            "columns": column_info,
            "total_rows": len(current_dataset),
            "total_columns": len(current_dataset.columns)
        }, indent=2)
        
    except Exception as e:
        return json.dumps({"error": str(e)})

def create_agent(api_key: str):
    """Create LangChain agent with tools"""
    llm = ChatGoogleGenerativeAI(
        model="gemini-2.0-flash-exp",
        google_api_key=api_key,
        temperature=0.3
    )
    
    tools = [perform_eda, forecast_data, get_column_info]
    
    prompt = ChatPromptTemplate.from_messages([
        ("system", """You are an expert data analyst assistant. You have access to tools for:
1. Exploratory Data Analysis (EDA) - analyze data distributions, correlations, and statistics
2. Forecasting - predict future values using time series models
3. Column Information - get detailed information about dataset columns

When a user asks to analyze data:
- Use perform_eda tool for exploratory analysis, statistical summaries, correlations, and visualizations
- Use forecast_data tool for time series predictions
- Use get_column_info tool to understand the dataset structure

Always provide clear, detailed explanations of your findings. When tools return visualizations, describe what they show.
Be proactive in suggesting relevant analyses based on the data characteristics."""),
        ("placeholder", "{chat_history}"),
        ("human", "{input}"),
        ("placeholder", "{agent_scratchpad}"),
    ])
    
    agent = create_tool_calling_agent(llm, tools, prompt)
    agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True, handle_parsing_errors=True)
    
    return agent_executor

def stream_chat(api_key: str, message: str, dataset_file: str = None, chat_history: List[Dict[str, str]] = None):
    """Stream chat responses"""
    if chat_history is None:
        chat_history = []
    
    # Load dataset if provided
    if dataset_file and dataset_file != "":
        load_result = load_dataset(dataset_file)
        if "error" in load_result:
            yield json.dumps({"type": "error", "message": load_result["error"]}) + "\n"
            return
    
    agent = create_agent(api_key)
    
    # Convert chat history to LangChain format
    formatted_history = []
    for msg in chat_history:
        if msg["role"] == "user":
            formatted_history.append(("human", msg["content"]))
        elif msg["role"] == "assistant":
            formatted_history.append(("ai", msg["content"]))
    
    try:
        # Stream the response
        for chunk in agent.stream({
            "input": message,
            "chat_history": formatted_history
        }):
            # Yield different types of chunks
            if "output" in chunk:
                yield json.dumps({"type": "content", "data": chunk["output"]}) + "\n"
            elif "intermediate_steps" in chunk:
                for step in chunk["intermediate_steps"]:
                    tool_name = step[0].tool
                    tool_input = step[0].tool_input
                    tool_output = step[1]
                    
                    yield json.dumps({
                        "type": "tool_call",
                        "tool": tool_name,
                        "input": tool_input,
                        "output": tool_output
                    }) + "\n"
    except Exception as e:
        yield json.dumps({"type": "error", "message": str(e)}) + "\n"

if __name__ == "__main__":
    # Command line interface for testing
    if len(sys.argv) < 2:
        print("Usage: python ai_tools.py <command> [args...]")
        sys.exit(1)
    
    command = sys.argv[1]
    
    if command == "load":
        if len(sys.argv) < 3:
            print("Usage: python ai_tools.py load <file_path>")
            sys.exit(1)
        result = load_dataset(sys.argv[2])
        print(json.dumps(result, indent=2))
    
    elif command == "chat":
        if len(sys.argv) < 4:
            print("Usage: python ai_tools.py chat <api_key> <message> [dataset_path]")
            sys.exit(1)
        api_key = sys.argv[2]
        message = sys.argv[3]
        dataset_file = sys.argv[4] if len(sys.argv) > 4 else None
        
        for chunk in stream_chat(api_key, message, dataset_file):
            print(chunk, end='', flush=True)
    
    else:
        print(f"Unknown command: {command}")
        sys.exit(1)
